walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,205 @@
+      emptySlots.add(TimeRange.fromStartEnd(0, END_OF_DAY, false));
+      return emptySlots; 
+    }
+
+    // A for loop is necessary because each consecutive pair 
+    // of time ranges needs to be looked at at each iteration. 
+    for (int i = 0; i < allBusyTimeslots.size(); i++ ) {
+
+      if (i == 0) {
+        int earliestStart = allBusyTimeslots.get(i).start();
+        if (earliestStart >= meetingDuration) {
+          emptySlots.add(TimeRange.fromStartEnd(START_OF_DAY, earliestStart, false)); 
+        }
+      } 
+
+      if (i == allBusyTimeslots.size()-1) {
+        int latestEnd = allBusyTimeslots.get(i).end();
+        if ((END_OF_DAY - latestEnd) >= meetingDuration) {
+          emptySlots.add(TimeRange.fromStartEnd(latestEnd, END_OF_DAY, false)); 
+        }
+      } else {
+        int currEnd = allBusyTimeslots.get(i).end();
+        int nextStart = allBusyTimeslots.get(i+1).start(); 
+
+        if (nextStart - currEnd >= meetingDuration) {
+          emptySlots.add(TimeRange.fromStartEnd(currEnd, nextStart, false)); 
+        }
+      }
+    } 
+    return emptySlots; 
+  }
+
+  /** Finds the common time range between time ranges of two collections */ 
+  private List<TimeRange> findCommonEmptySlots(
+      List<TimeRange> mandatoryAttendeesEmptySlots, List<TimeRange> optionalAttendeesEmptySlots, long meetingDuration) {
+        
+      // If any of mandatory or optional attendee has no free time, 
+      // return the other group's free time slots. 
+      if (mandatoryAttendeesEmptySlots.size() == 0) {
+        return optionalAttendeesEmptySlots;
+      }
+      if (optionalAttendeesEmptySlots.size() == 0) {
+        return mandatoryAttendeesEmptySlots; 
+      }    
+
+      List<TimeRange> commonSlots = new ArrayList<TimeRange>();
+      int i = 0;
+      int j = 0;
+
+      while (i < mandatoryAttendeesEmptySlots.size() && j < optionalAttendeesEmptySlots.size()) {
+        TimeRange mandatorySlot = mandatoryAttendeesEmptySlots.get(i);
+        TimeRange optionalSlot = optionalAttendeesEmptySlots.get(j);
+
+        // If two time ranges overlap, find the intersection of these two. 
+        if (mandatorySlot.overlaps(optionalSlot)) {
+          int start = Math.max(mandatorySlot.start(), optionalSlot.start());
+          int end = Math.min(mandatorySlot.end(), optionalSlot.end());
+          
+          if (end - start >= meetingDuration) {
+            TimeRange commonSlot = TimeRange.fromStartEnd(start, end, false);
+            commonSlots.add(commonSlot);
+          } else {
+            commonSlots.add(mandatorySlot); 
+          }
+        } 
+
+        // Move the pointer pointing at the time range that ends earlier,
+        // because the time range that ends later might have intersection with other
+        // time ranges in the time ranges taht come after the one that ends earlier.
+        if (mandatorySlot.end() >= optionalSlot.end()) {
+          j ++; 
```

Comment:
We don't need the space before ++. Here and below.

Reply:
Updated.